runtime/src/main/java/org/corfudb/util/serializer/KeyDynamicProtobufSerializer.java

```diff
@@ -0,0 +1,135 @@
+public class KeyDynamicProtobufSerializer extends DynamicProtobufSerializer {
+
+    public KeyDynamicProtobufSerializer(CorfuRuntime corfuRuntime) {
+        super(corfuRuntime);
+    }
+
+    /**
+     * Deserialize an object from a given byte buffer.
+     *
+     * @param b The bytebuf to deserialize.
+     * @return The deserialized object.
+     */
+    @Override
+    public Object deserialize(ByteBuf b, CorfuRuntime rt) {
+
+        try (ByteBufInputStream bbis = new ByteBufInputStream(b)) {
+            ProtobufSerializer.MessageType type = ProtobufSerializer.MessageType.valueOf(bbis.readInt());
+            int size = bbis.readInt();
+            byte[] data = new byte[size];
+            bbis.readFully(data);
+            CorfuStoreMetadata.Record corfuRecord = CorfuStoreMetadata.Record.parseFrom(data);
+            Any payload = corfuRecord.getPayload();
+
+            String fullMessageName = getFullMessageName(payload);
+            if (!messagesFdProtoNameMap.containsKey(fullMessageName)) {
+                log.error("messagesFdProtoNameMap doesn't contain the message type {} of payload {}." +
+                                "Please check if the related table is properly opened with correct schema.",
+                        fullMessageName, payload);
+                log.error("messagesFdProtoNameMap keySet is {}", messagesFdProtoNameMap.keySet());
+            }
+
+            if (type.equals(ProtobufSerializer.MessageType.KEY)) {
+                Descriptors.FileDescriptor valueFileDescriptor
+                        = getDescriptor(messagesFdProtoNameMap.get(fullMessageName));
+                Descriptors.Descriptor valueDescriptor = valueFileDescriptor.findMessageTypeByName(getMessageName(payload));
+                DynamicMessage value = DynamicMessage.parseFrom(valueDescriptor, payload.getValue());
+                return new CorfuDynamicKey(payload.getTypeUrl(), value);
+            }
+
+            String metadataTypeUrl = null;
+            ByteString metaDataByteString = null;
+            if (corfuRecord.hasMetadata()) {
+                Any anyMetadata = corfuRecord.getMetadata();
+                metadataTypeUrl = anyMetadata.getTypeUrl();
+                metaDataByteString = anyMetadata.getValue();
+
+                String fullMetadataMessageName = getFullMessageName(anyMetadata);
+                if (!messagesFdProtoNameMap.containsKey(fullMetadataMessageName)) {
+                    log.error("messagesFdProtoNameMap doesn't contain the message type {} of metadata {}." +
+                                    "Please check if the related table is properly opened with correct schema.",
+                            fullMetadataMessageName, anyMetadata);
+                    log.error("messagesFdProtoNameMap keySet is {}", messagesFdProtoNameMap.keySet());
+                }
+            }
+            return new OpaqueCorfuDynamicRecord(payload.getTypeUrl(), payload.getValue(), metadataTypeUrl, metaDataByteString);
```

Comment:
did you miss deserializing the metadata from the descriptor obtained from the map?  Otherwise what is the use of the 'if' check above?

Reply:
This serializer will only serialize / deserialize the Key. The if-check above serves the purpose of validating the protobuf types. If someone reopened a table with a wrong schema, it overwrites the protobuf type. In that case the periodic compactor will capture this issue and it will log the error here. Let me add a comment to the if-check to improve readability.